File: src/DO/Graphics/DerivedQObjects/UserThread.cpp
// ========================================================================== //
// This file is part of DO++, a basic set of libraries in C++ for computer 
// vision.
//
// Copyright (C) 2013 David Ok <david.ok8@gmail.com>
//
// This Source Code Form is subject to the terms of the Mozilla Public 
// License v. 2.0. If a copy of the MPL was not distributed with this file, 
// you can obtain one at http://mozilla.org/MPL/2.0/.
// ========================================================================== //

#include "../GraphicsUtilities.hpp"
#include "UserThread.hpp"
#include <QDebug>

namespace DO {

  UserThread::UserThread(QObject* parent)
    : QThread(parent)
    , userMain_(0)
    , doWaitForClick_(false)
    , doWaitForKey_(false)
  {
  }

  int UserThread::getMouse(int& x, int& y)
  {
    mutex_.lock();
    doWaitForClick_ = true;
    condition_.wait(&mutex_);
    mutex_.unlock();
    x = mouseX_; y = mouseY_;
    if (mouseButton_ == Qt::LeftButton)
      return MOUSE_LEFT_BUTTON;
    else if (mouseButton_ == Qt::MiddleButton)
      return MOUSE_MIDDLE_BUTTON;
    else if (mouseButton_ == Qt::RightButton)
      return MOUSE_RIGHT_BUTTON;
    else
      return MOUSE_NO_BUTTON;
  }

  int UserThread::getKey()
  {
    mutex_.lock();
    doWaitForKey_ = true;
    condition_.wait(&mutex_);
    mutex_.unlock();
    return key_;
  }

  void UserThread::listenToWindowEvents()
  {
    mutex_.lock();
    condition_.wait(&mutex_);
    mutex_.unlock();
  }

  void UserThread::getEvent(Event& e)
  {
    mutex_.lock();
    e = event_;
    mutex_.unlock();
  }

  void UserThread::pressedMouseButtons(int x, int y, Qt::MouseButtons buttons)
  {
    mutex_.lock();
    if (doWaitForClick_)
    {
      doWaitForClick_ = false;
      mouseButton_ = buttons; mouseX_ = x; mouseY_ = y;
      condition_.wakeOne();
    }
    mutex_.unlock();
  }

  void UserThread::pressedKey(int key)
  {
    mutex_.lock();
    if (doWaitForKey_)
    {
      doWaitForKey_ = false;
      key_ = key;
      condition_.wakeOne();
    }
    mutex_.unlock();
  }

  void UserThread::receivedEvent(Event e)
  {
    mutex_.lock();
    condition_.wakeOne();
    event_ = e;
    mutex_.unlock();
  }

  void UserThread::closedWindow()
  {
    // What to do in case the window was closed by the user and that 
    // the program was expecting for a mouse click or a pressed key?
    mutex_.lock();
    if (doWaitForClick_ || doWaitForKey_)
    {
      qDebug() << "Window closed unexpectedly while waiting for user input!";
      condition_.wakeOne();
    }
    //qDebug() << "Closing window as requested!";
    mutex_.unlock();
  }

  void UserThread::run()
  {
    //qDebug("Running user thread!");
    if (!userMain_)
    {
      qFatal("No registered user main!");
      return;
    }
    int ret;
    ret = userMain_(getGuiApp()->argc, getGuiApp()->argv);
    return;
  }

} /* namespace DO */

File: python/do/sara/graphics/derived_qobjects/user_thread.py
from PySide2.QtCore import QMutex, QObject, QThread, Qt, QWaitCondition, Signal
from PySide2.QtWidgets import QApplication


class UserThreadSignals(QObject):
    create_window = Signal(int, int)
    draw_point = Signal(int, int, object)
    draw_image = Signal(object, object, float)


class UserThread(QThread):

    def __init__(self, parent=None):
        super(UserThread, self).__init__(parent)
        self._user_main = None
        self._mutex = QMutex()
        self._condition = QWaitCondition()

        self._do_wait_for_click = False
        self._mouse_button = Qt.MouseButtons
        self._mouse_x = None
        self._mouse_y = None

        self._do_wait_for_key = False
        self._key = None

        self._event = None

        self.signals = UserThreadSignals()
        self.finished.connect(QApplication.instance().quit)

    def register_user_main(self, user_main):
        self._user_main = user_main

    def get_mouse(self):
        pass

    def get_key(self):
        self._mutex.lock()
        self._do_wait_for_key = True
        self._condition.wait(self._mutex)
        self._mutex.unlock()
        return self._key

    def pressed_key(self, key):
        self._mutex.lock()
        if self._do_wait_for_key:
            self._do_wait_for_key = False
            self._key = key
            self._condition.wakeOne()
        self._mutex.unlock()

    @property
    def key(self):
        return self._key

    def get_event(self, event):
        pass

    def listenToWindowEvents(self):
        pass

    def pressedMouseButtons(self, x, y, buttons):
        pass

    def closedWindow(self):
        pass

    def receivedEvent(self, event):
        pass

    def sendEvent(self, event, delayMs):
        pass

    def run(self):
        if self._user_main is None:
            return
        self._user_main()
